azure-toolkit-libs/azure-toolkit-common-lib/src/main/java/com/microsoft/azure/toolkit/lib/common/operation/AzureOperationAspect.java

```diff
@@ -28,38 +28,31 @@ public void operation() {
 
     @Before("operation()")
     public void beforeEnter(JoinPoint point) {
-        enterOperation(point);
+        final AzureOperationRef operation = toOperationRef(point);
+        AzureTelemeter.beforeEnter(operation);
+        AzureTaskContext.current().pushOperation(operation);
     }
 
     @AfterReturning("operation()")
     public void afterReturning(JoinPoint point) {
-        exitOperation(point);
+        final AzureOperationRef current = toOperationRef(point);
+        final AzureOperationRef operation = (AzureOperationRef) AzureTaskContext.current().popOperation();
+        AzureTelemeter.afterExit(operation);
+        assert Objects.equals(current, operation) : String.format("popped operation[%s] is not the exiting operation[%s]", current, operation);
     }
 
     @AfterThrowing(pointcut = "operation()", throwing = "e")
     public void afterThrowing(JoinPoint point, Throwable e) throws Throwable {
-        final AzureOperationRef operation = exitOperation(point);
+        final AzureOperationRef current = toOperationRef(point);
```

Comment:
the validation seems duplicate with `afterReturning` (except the assert message), can we share the logic?

Reply:
no, actually there are only two-lines same code. sharing code will make the code ugly.